### PR TITLE
Booking form

### PR DESCRIPTION
**Pull request by simonpizevski**
I have created a form for submitting a booking to our database in the collection "bookings". The collection includes: id, email, fullName, movieTitle, movieTime and screeningId. 

Skärmbild 2024-05-16 112334

When the user presses the "boka" button the booking form opens at "/boka-film" with params for title, screeningId and time.  

Skärmbild 2024-05-16 145826

This is an example of the form. And when a successfull booking is submitted a modal pops up to show that is was successfull. 

Skärmbild 2024-05-16 112313

### CODE REVIEW

File: app/boka-film/page.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+'use client';
+
+import BookingForm from '@/app/components/BookingForm';
+import { useSearchParams } from 'next/navigation';
+import React from 'react';
+
+const BookMoviePage: React.FC = () => {
+  const searchParams = useSearchParams();
```
Comment (NelsonTheBrave) [inline app/boka-film/page.tsx]: I think there is a problem with this way of getting the information about a screening that you are doing. You can actually create fake screenings by altering the URL manually and then booking it to our database. The way to avoid this would be to only add the screenings _id as a search parameter and then fetch the relevant information from the database inside you boka-film/page.tsx. I also think it would make more sense to get the poster information in that same database fetch instead of fetching it already in the screenings component on the startpage like you have done now, since the poster is not used there. Feel like that makes the code clearer and maybe improves its separation of concerns.

image

File: app/lib/definitions.ts [hunk 2]
```diff
@@ -19,8 +19,10 @@ export type carouselSettings = {
 
 //This type is based on the info in our database, at some point we should fix so we don't have both Screenings and Screening as types if it doesn't fill a purpose
 export type Screening = {
+  _id: number;
   movie: string;
   saloon: string;
   date: string;
   runtime?: string;
+  poster: string;
```
Comment (NelsonTheBrave) [inline app/lib/definitions.ts]: I don't think "poster" is necessarily a property on all screening objects, for example it will not be used when presenting specific screenings on the /filmer/[id] page, so I think it should have a "?" like runtime

File: app/ui/screeningsTableStartpage.tsx [hunk 3]
```diff
@@ -13,20 +14,27 @@ import { DatePicker } from '@mui/x-date-pickers/DatePicker';
...
+  const [selectedDate, setSelectedDate] = React.useState<Dayjs | null>(
+    dayjs(new Date())
+  );
 
   const [screenings, setScreenings] = useState<Screening[]>([]);
 
   const fetchScreenings = async (URL: string) => {
     const res = await fetch(URL);
     const screenings = await res.json();
+    console.log('Screenings:', screenings);
```
Comment (NelsonTheBrave) [inline app/ui/screeningsTableStartpage.tsx]: get rid of the console log :)

File: app/om-oss/page.tsx
Comment (NelsonTheBrave) [inline app/om-oss/page.tsx:1]: Why did you remove the (pages) folder? Or maybe you just didn't have an updated main that you were building your branch on? Anyways I think we should keep the (pages) folder because it makes the folder structure clearer

File: app/ui/screeningsTableStartpage.tsx [hunk 4]
```diff
@@ -13,20 +14,27 @@ import { DatePicker } from '@mui/x-date-pickers/DatePicker';
 import dayjs, { Dayjs } from 'dayjs';
 import 'dayjs/locale/de';
 import { Screening } from '../lib/definitions';
+import { useSearchParams } from 'next/navigation';
```
Comment (NelsonTheBrave) [inline app/ui/screeningsTableStartpage.tsx]: You're importing useSearchParams and later assigning it's function to a constant, but you never use it so I think you should take it away.

**Review by NelsonTheBrave — CHANGES_REQUESTED** [review]
Overall the bookings page looks really nice and the booking functionality works, good job :)

I have already written somethings for you on discord but I add them here as well:
- The time shown for a movie screening on the bookings page is two hours ahead of the actual screening time according to our database
- The name field in booking requires first name and last name but there is no info about that (except for an error message), so I think it would be nice to add e.g. a placeholder communicating this
- On my screen the buttons ended up being inside the footer as in the picture:
Screenshot 2024-05-16 154640

Except for these things I think the biggest issue, as I've commented in the code, is that you can create non-existant screenings by manually altering the search parameters in the /boka-film-URL. I have given a suggestion on how change fix this in the code comment.

File: app/lib/definitions.ts (hunk 2 above)
Comment (simonpizevski) [reply app/lib/definitions.ts]: Good note! Ive changed it now.

File: app/ui/screeningsTableStartpage.tsx (hunk 4 above)
Comment (simonpizevski) [reply app/ui/screeningsTableStartpage.tsx]: You are right, ive removed it now.

**Review by JrSchoolAccount — APPROVED** [review]
Approved as we discussed in the meeting.
And we will fix the minor issues in another PR .

**Review by NelsonTheBrave — APPROVED** [review]
As Jesper wrote above, let's approve this to include it in main and fix the issue with fake bookings soon

**Comment by JrSchoolAccount** [comment]
Managed to fix all but one error:
```
Error occurred prerendering page "/boka-film". Read more: https://nextjs.org/docs/messages/prerender-error

   Generating static pages (4/18)  [=   ]Error fetching movies: TypeError: Failed to parse URL from /api/movies
    at new Request (node:internal/deps/undici/undici:7152:19)
    at a (/Users/maclaptop/Desktop/Vscode/kino-site-nextjs/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:180691)
    ... 2 lines matching cause stack trace ...
    at async i (/Users/maclaptop/Desktop/Vscode/kino-site-nextjs/.next/server/app/page.js:3:140741) {
  [cause]: TypeError [ERR_INVALID_URL]: Invalid URL
      at new NodeError (node:internal/errors:405:5)
      at new URL (node:internal/url:778:13)
      at new Request (node:internal/deps/undici/undici:7150:25)
      at a (/Users/maclaptop/Desktop/Vscode/kino-site-nextjs/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:12:180691)
      at B (/Users/maclaptop/Desktop/Vscode/kino-site-nextjs/.next/server/chunks/948.js:2:6974)
      at /Users/maclaptop/Desktop/Vscode/kino-site-nextjs/.next/server/chunks/948.js:2:9220
      at async i (/Users/maclaptop/Desktop/Vscode/kino-site-nextjs/.next/server/app/page.js:3:140741) {
    input: '/api/movies',
    code: 'ERR_INVALID_URL'
```
